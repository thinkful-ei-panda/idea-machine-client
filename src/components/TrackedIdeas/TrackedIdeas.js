import React from 'react'
import Results from '../IdeaResults/IdeaResults'
import config from '../../config'
import TokenService from '../../services/token-service'

class TrackedIdeas extends React.Component{
  state = {
    results: [],
    error:null,
    loading:false,
  }

  componentDidMount(){
    this.setState({loading:true})
    fetch(`${config.API_ENDPOINT}/followedIdeas`,{
      method: 'GET',
      headers: {
        'content-type':'application/json',
        'Authorization':`bearer ${TokenService.getAuthToken()}`
      }
    })
    .then(res => (!res.ok)
    ?res.json().then(e => Promise.reject(e))
    :res.json())
    .then(results => {
      this.setState({results,loading:false})
    })
    .catch(error => {
      this.setState({error,loading:false})
    })
  }

  handleUnfollowClick = (e) => {
    this.setState({loading:true})
    const idea_id = e.target.closest('li').id

    fetch(`${config.API_ENDPOINT}/followedIdeas/${idea_id}`,{
      method: 'DELETE',
      headers: {
        'content-type':'application/json',
        'Authorization':`bearer ${TokenService.getAuthToken()}`,
      },
      body:JSON.stringify({
        follower_id:TokenService.getPayloadFromToken.user_id
      })
    })
    .then(() => {
      const newResults = this.state.results.filter(result => result.id !== Number(idea_id))
      this.setState({results:newResults,loading:false})
    })
    .catch(error => {
      this.setState({error,loading:false})
    })
  }

  render(){
    const {results} = this.state
    return (
      <>        
        {this.state.error && <div className='error'>{this.state.error.error}</div>}
        {/* {this.state.loading && <div className='loading'>Loading...</div>} */}
        {results.length !== 0 && <Results 
        handleUnfollowClick = {this.handleUnfollowClick}
        loading={this.state.loading}
        results={results} />}
      </>
    )
  }
}

export default TrackedIdeas