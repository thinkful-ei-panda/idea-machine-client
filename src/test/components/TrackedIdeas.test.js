import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import TrackedIdeas from '../../components/TrackedIdeas/TrackedIdeas';
describe(`TrackedIdeas Component`, () => {
    describe(`Smoke test`, () => {
        it(`Renders without crashing`, () => {
            const div = document.createElement('div');
            ReactDOM.render(
                <BrowserRouter>
                    <TrackedIdeas />
                </BrowserRouter>
                , div);
            ReactDOM.unmountComponentAtNode(div);
        });
    });
    
    describe(`Snapshot test`, () => {
        it(`Renders the UI as expected`, () => {
            const tree = renderer
                .create(
                    <BrowserRouter>
                        <TrackedIdeas />
                    </BrowserRouter>
                )
                .toJSON()
            expect(tree).toMatchSnapshot()
        });
    });
})