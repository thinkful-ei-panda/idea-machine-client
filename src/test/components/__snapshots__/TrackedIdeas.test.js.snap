// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackedIdeas Component Snapshot test Renders the UI as expected 1`] = `null`;
